docs/RationaleSymbols.md

```diff
@@ -0,0 +1,79 @@
+are analogous in scoping and structure to functions and call instructions.  The 
+top-level mlir module or firrtl circuit op define a symbol table and all 
+modules contained define symbols, with instances refering by symbol to their 
+instantiated module.
+
+## Inner Symbol
+
+Within a module (firrtl or hw), many entities may exist which can be referenced 
+outside the module.  Operations and ports (and memory ports), need to define 
+symbol-like data to allow forming non-ssa linkage between disparate elements.  
+To accomplish this, an attribute `inner_sym` is created providing a scoped 
+symbol-like name to the element.  An `inner_sym` operation resides in 
+arbitrarily-nested regions of a region that defines an `InnerSymbolTable` and
+an `Symbol` .
+
+Inner Symbols are different from normal symbols due to MLIR symbol table 
+resolution rules.  Specifically normal symbols are resolved by first going up 
+to the closest parent symbol table and resolving down from there (recursing 
+down for complex symbol paths).  In FIRRTL and SV, modules define a symbol in a 
+circuit or std.module symbol table.  For instances to be able to resolve the 
+modules they instantiate, the symbol use in an instance must resolve in the 
+top-level symbol table.  If a module were a symbol table, instances resolving a 
+symbol would start from their own module, never seeing other modules (since 
+resolution would start in the parent module of the instance and be unable to go 
+to the global scope).  The second problem arises from nesting.  Symbols 
+defining ops must be immediate children of a symbol table.  FIRRTL and SV 
+operations which define a inner_sym are grandchildren, at least, of a symbol 
+table and may be much further nested.  Lastly, ports need to define inner_sym, 
+something not allowed by normal symbols.
+
+## Inner Symbol Reference Attribute
+
+An attribute `InnerRefAttr` is provided to encapsulated references to inner 
+symbols.  This attribute stores the parent symbol and the inner symbol.  This 
+provides a uniform type for storing and manipulating references to inner 
+symbols.
+
+## Cost
+
+Inner symbols are more costly than normal symbols, precisely from the 
+relaxation of mlir symbol constraints.  Since nested regions are allows, 
+finding all ops defining an inner name requires a recursive ir scan.  
+Verification is likewise trickier, partly due the significant increase in 
+non-local references.
+
+## Common Use
+
+The most common use for InnerRefs are to build paths through the instantiation 
+graph to use a subsetset of the instances of an entitity is some way.  This may 
+be reading values via System Verilog XMRs, specifying SV bind constraints, 
+specifying placement constraints, or representing non-local attributes (FIRRTL).
+
+The commone element for building paths of instances through the instantiation 
+graph is with a `NameRefArrayAttr` attribute.  This is used, for example, by 
+`sv.GlobalRefOp` and `firrtl.NonLocalAnchor` .
```

Comment:
```suggestion
The most common use for `InnerRefAttr`s are to build paths through the instantiation 
graph to use a subset of the instances of an entity is some way.  This may 
be reading values via System Verilog cross-module references (XMRs), specifying SV bind constraints, 
specifying placement constraints, or representing non-local attributes (FIRRTL).

The common element for building paths of instances through the instantiation 
graph is with a `NameRefArrayAttr` attribute.  This is used, for example, by 
`sv.GlobalRefOp` and `firrtl.NonLocalAnchor` .
```